pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
 	if err := LoadBpfObjects(&objs, &ebpf.CollectionOptions{Maps: ebpf.MapOptions{PinPath: pinDir}}); err != nil {
 		return nil, fmt.Errorf("loading objects: pinDir:%s, err:%s", pinDir, err)
 	}
-	return &IngNodeFwController{
+	infc := &IngNodeFwController{
 		objs:    objs,
 		pinPath: pinDir,
 		links:   make(map[string]link.Link, 0),
-	}, nil
+	}
+	// Load pinned links from /sys/fs/bpf/xdp_ingress_node_firewall_process on initialization.
+	// That way, the state in /sys/fs/bpf/xdp_ingress_node_firewall_process and the tracked list of links
+	// will be in sync.
+	if err := infc.loadPinnedLinks(); err != nil {
+		return nil, err
+	}
+
+	return infc, nil
 }
 
-// IngressNodeFwRulesLoader Add/Update/Delete ingress nod firewll rules to eBPF LPM MAP
-func (infc *IngNodeFwController) IngressNodeFwRulesLoader(ingFireWallConfig ingressnodefwiov1alpha1.IngressNodeFirewallRules,
-	isDelete bool, ifId uint32) error {
-	objs := infc.objs
-	info, err := objs.BpfMaps.IngressNodeFirewallTableMap.Info()
+// IngressNodeFwRulesLoader adds/updates/deletes ingress node firewall rules to the eBPF LPM MAP in an idempotent way.
+// IngressNodeFwRulesLoader executes the following actions in order:
+// i)   Get eBPF objs to create/update eBPF maps and get map info.
+// ii)  Build a map of valid ebpfKeys pointing to the ebpfRules that should be associated to them (built from
+//      ifaceIngressRules).
+// iii) Get stale keys (= keys inside the eBPF map but not inside the currently desired ruleset).
+// iv)  Purge all stale keys from the eBPF map.
+// v)   Add/update all keys. This is an idempotent action and non-existing keys are added whereas existing keys
+//      are updated.
+// vi)  Generate ingress node firewall events.
+// In the context of this method, stale keys are keys that figure inside the eBPF map but that are not generated
+// during step ii) from the provided ingressRules slice.
+func (infc *IngNodeFwController) IngressNodeFwRulesLoader(
+	ifaceIngressRules map[string][]v1alpha1.IngressNodeFirewallRules) error {
+	// Get eBPF objs to create/update eBPF maps and get map info.
+	info, err := infc.objs.BpfMaps.IngressNodeFirewallTableMap.Info()
+	if err != nil {
+		return fmt.Errorf("cannot get map info: %v", err)
+	}
+	klog.Infof("Ingress node firewall map Info: %+v with FD %s", info, infc.objs.BpfMaps.IngressNodeFirewallTableMap.String())
+
+	// Convert IngressNodeFirewallRules into data that can be written to the BPF map.
+	// Build a map of valid ebpfKeys pointing to the ebpfRules that should be associated to them.
+	ebpfKeyToRules := make(map[BpfLpmIpKeySt]BpfRulesValSt)
+	for interfaceName, ingressRules := range ifaceIngressRules {
+		// Look up the network interface by name.
```

Comment:
use the new method in interfaces `GetInterfaceIndex`

Reply:
done